File: java/src/main/java/com/mertsy/javasample/MainActivity.java
package com.mertsy.javasample;

import android.Manifest;
import android.annotation.SuppressLint;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.os.Parcelable;
import android.util.Log;
import android.view.View;
import android.widget.LinearLayout;

import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatButton;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.appcompat.widget.SwitchCompat;
import androidx.core.content.ContextCompat;

import com.mertsy.capturing.CapturingType;
import com.mertsy.capturing.MertsyCapturing;
import com.mertsy.capturing.MertsyCapturingSession;
import com.mertsy.capturing.ModelAccessType;
import com.mertsy.common.TokenStatus;
import com.mertsy.sdk.MertsySDK;

public class MainActivity extends BaseActivity {

    private static final String TAG = "MainActivity";

    private boolean hasStoragePermission() {
        return isPermissionGranted(targetStoragePermission());
    }

    private boolean hasCameraPermission() {
        return isPermissionGranted(Manifest.permission.CAMERA);
    }

    private String targetStoragePermission() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU ?
                Manifest.permission.READ_MEDIA_IMAGES :
                Manifest.permission.WRITE_EXTERNAL_STORAGE;
    }

    private boolean isPermissionGranted(String permission) {
        return ContextCompat.checkSelfPermission(
                this,
                permission
        ) == PackageManager.PERMISSION_GRANTED;
    }

    private boolean isTokenValid() {

        switch (MertsySDK.getTokenStatus()) {
            case CHECKING: {
                String msg = "Token validation. Try again later";
                Log.e(TAG, msg);
                showToast(msg);
                return false;
            }
            case ERROR: {
                String msg = "Wrong token";
                Log.e(TAG, msg);
                showToast(msg);
                return false;
            }
            default:
                return true;
        }
    }

    @Override
    @SuppressLint("SetTextI18n")
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        AppCompatTextView tvTokenStatus = findViewById(R.id.tvTokenStatus);
        AppCompatTextView tvFrameworkVersion = findViewById(R.id.tvFrameworkVersion);

        SwitchCompat switchCacheSession = findViewById(R.id.switchCacheSession);
        SwitchCompat switchModelAccess = findViewById(R.id.switchModelAccess);

        AppCompatButton btnPanorama = findViewById(R.id.btnPanorama);
        AppCompatButton btnIndoor = findViewById(R.id.btnIndoor);
        AppCompatButton btnCars = findViewById(R.id.btnCars);
        AppCompatButton btnView = findViewById(R.id.btnView);

        //Required for capturing
        requestCameraAndStoragePermissions();

        tvTokenStatus.setText("Token status: Checking");

        MertsySDK.setOnTokenStatusChangedListener(tokenStatus -> {
            String status = "Token status: ";
            if (tokenStatus == TokenStatus.SUCCESS) {
                status += "Success";
            } else {
                status += "Error";
            }

            tvTokenStatus.setText(status);
        });

        tvFrameworkVersion.setText("SDK Version: " + MertsySDK.getVersion());

        switchCacheSession.setChecked(MertsyCapturing.getCacheSession());
        switchCacheSession.setOnCheckedChangeListener((buttonView, isChecked) -> MertsyCapturing.setCacheSession(isChecked));

        switchModelAccess.setChecked(MertsyCapturing.getModelAccess() == ModelAccessType.UNLISTED);
        switchModelAccess.setOnCheckedChangeListener((buttonView, isChecked) -> MertsyCapturing.setModelAccess(isChecked ? ModelAccessType.UNLISTED : ModelAccessType.PUBLIC));

        btnIndoor.setOnClickListener(v -> startCapturingActivity(CapturingType.INDOOR));
        btnCars.setOnClickListener(v -> startCapturingActivity(CapturingType.CARS_AND_VEHICLES));
        btnPanorama.setOnClickListener(v -> startCapturingActivity(CapturingType.PANORAMA));

        btnView.setOnClickListener(v -> {
            if (isTokenValid()) {
                startActivity(new Intent(this, ModelViewActivity.class));
            }
        });

    }

    @Override
    protected void onStart() {
        super.onStart();

        //Ask user to restore his session
        MertsyCapturingSession session = MertsyCapturing.getSession();
        if (session != null) {
            LinearLayout llPreviousSession = findViewById(R.id.llPreviousSession);
            AppCompatImageView ivPreviousSession = findViewById(R.id.ivPreviousSession);
            AppCompatButton btnResume = findViewById(R.id.btnResume);
            AppCompatButton btnDelete = findViewById(R.id.btnDelete);

            llPreviousSession.setVisibility(View.VISIBLE);
            ivPreviousSession.setImageBitmap(session.getPreview());

            //We will open the last session, so type is not matter
            btnResume.setOnClickListener(v -> startCapturingActivity(null));

            btnDelete.setOnClickListener(v -> {
                MertsyCapturing.clearSessionCaches();
                llPreviousSession.setVisibility(View.GONE);
            });
        }

    }

    private void startCapturingActivity(@Nullable CapturingType type) {
        if (!isTokenValid())
            return;

        //<3 gb free
        if (!MertsyCapturing.getHasEnoughMemory()) {
            String msg = "Not enough memory, the work might be unstable";
            Log.e(TAG, msg);
            showToast(msg);
        }

        if (!MertsyCapturing.getHasRequiredSensors()) {
            String msg = "No gyroscope, the capturing is impossible";
            Log.e(TAG, msg);
            showToast(msg);
            return;
        }

        Intent capturingIntent = new Intent(this, CapturingActivity.class);
        if (type != null) {
            capturingIntent.putExtra(CapturingActivity.TAG, (Parcelable) type);
        }

        startActivity(capturingIntent);
    }

    private void requestCameraAndStoragePermissions() {
        String[] permissions = new String[]{Manifest.permission.CAMERA, targetStoragePermission()};

        if (!hasStoragePermission() || !hasCameraPermission())
            requestPermissions(permissions, 123);
    }
}

File: kotlin/src/main/java/com/mertsy/kotlinsample/MainActivity.kt
package com.mertsy.kotlinsample

import android.Manifest
import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.os.Parcelable
import android.util.Log
import android.widget.LinearLayout
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.AppCompatButton
import androidx.appcompat.widget.AppCompatImageView
import androidx.appcompat.widget.AppCompatTextView
import androidx.appcompat.widget.SwitchCompat
import androidx.core.content.ContextCompat
import androidx.core.view.isVisible
import com.mertsy.capturing.CapturingType
import com.mertsy.capturing.MertsyCapturing
import com.mertsy.capturing.ModelAccessType
import com.mertsy.common.TokenStatus
import com.mertsy.sdk.MertsySDK

class MainActivity : AppCompatActivity(R.layout.activity_main) {

    companion object {
        private const val TAG = "MainActivity"
    }

    private val Context.hasStoragePermission: Boolean
        get() = isPermissionGranted(
            targetStoragePermission
        )
    private val Context.hasCameraPermission get() = isPermissionGranted(Manifest.permission.CAMERA)

    private val targetStoragePermission
        get() = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            Manifest.permission.READ_MEDIA_IMAGES
        } else Manifest.permission.WRITE_EXTERNAL_STORAGE

    @SuppressLint("SetTextI18n")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        //Required for capturing
        requestCameraAndStoragePermissions()

        findViewById<AppCompatTextView>(R.id.tvTokenStatus).text = "Token status: Checking"

        MertsySDK.onTokenStatusChangedCallback = { status: TokenStatus ->
            val prefix = "Token status: "
            findViewById<AppCompatTextView>(R.id.tvTokenStatus).text = when (status) {
                TokenStatus.SUCCESS -> prefix + "Success"
                else -> prefix + "Error"
            }
        }

        findViewById<AppCompatTextView>(R.id.tvFrameworkVersion).text =
            "SDK Version: ${MertsySDK.getVersion()}"

        with(findViewById<SwitchCompat>(R.id.switchCacheSession)) {
            isChecked = MertsyCapturing.getCacheSession()
            setOnCheckedChangeListener { _, isEnabled ->
                MertsyCapturing.setCacheSession(isEnabled)
            }
        }

        with(findViewById<SwitchCompat>(R.id.switchModelAccess)) {
            isChecked = MertsyCapturing.modelAccess == ModelAccessType.UNLISTED
            setOnCheckedChangeListener { _, isEnabled ->
                MertsyCapturing.modelAccess =
                    isEnabled then ModelAccessType.UNLISTED ?: ModelAccessType.PUBLIC
            }
        }

        findViewById<AppCompatButton>(R.id.btnPanorama).setOnClickListener {
            startCapturingActivity(CapturingType.PANORAMA)
        }

        findViewById<AppCompatButton>(R.id.btnIndoor).setOnClickListener {
            startCapturingActivity(CapturingType.INDOOR)
        }

        findViewById<AppCompatButton>(R.id.btnCars).setOnClickListener {
            startCapturingActivity(CapturingType.CARS_AND_VEHICLES)
        }

        findViewById<AppCompatButton>(R.id.btnView).setOnClickListener {
            checkTokenAndExecute {
                startActivity(Intent(this, ModelViewActivity::class.java))
            }
        }
    }

    override fun onStart() {
        super.onStart()
        //Ask user to restore his session
        MertsyCapturing.session?.let { previousSession ->
            val llSessionPreview = findViewById<LinearLayout>(R.id.llPreviousSession)
            llSessionPreview.isVisible = true

            findViewById<AppCompatImageView>(R.id.ivPreviousSession).setImageBitmap(previousSession.preview)

            findViewById<AppCompatButton>(R.id.btnResume).setOnClickListener {
                //We will open the last session, so type is not matter
                startCapturingActivity(null)
            }

            findViewById<AppCompatButton>(R.id.btnDelete).setOnClickListener {
                MertsyCapturing.clearSessionCaches()
                llSessionPreview.isVisible = false
            }
        }
    }

    private fun checkTokenAndExecute(action: () -> Unit) {
        when (MertsySDK.tokenStatus) {
            TokenStatus.CHECKING -> {
                val msg = "Token validation. Try again later"
                Log.e(TAG, msg)
                showToast(msg)
            }

            TokenStatus.ERROR -> {
                val msg = "Wrong token"
                Log.e(TAG, msg)
                showToast(msg)
            }

            TokenStatus.SUCCESS -> action.invoke()
        }
    }

    private fun startCapturingActivity(type: CapturingType?) {
        checkTokenAndExecute {

            //<3 gb free
            if (!MertsyCapturing.hasEnoughMemory) {
                val msg = "Not enough memory, the work might be unstable"
                Log.e(TAG, msg)
                showToast(msg)
            }

            if (!MertsyCapturing.hasRequiredSensors) {
                val msg = "No gyroscope, the capturing is impossible"
                Log.e(TAG, msg)
                showToast(msg)
                return@checkTokenAndExecute
            }

            startActivity(
                Intent(this, CapturingActivity::class.java).apply {
                    type?.let { putExtra(CapturingActivity.TAG, it as Parcelable) }
                })
        }
    }

    private fun requestCameraAndStoragePermissions() {
        val permissions = arrayOf(
            targetStoragePermission,
            Manifest.permission.CAMERA
        )
        if (!hasStoragePermission || !hasCameraPermission)
            requestPermissions(permissions, 123)
    }

    private fun Context.isPermissionGranted(permission: String): Boolean {
        return ContextCompat.checkSelfPermission(
            this,
            permission
        ) == PackageManager.PERMISSION_GRANTED
    }
}